ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,199 @@
+    IndexSegment indexSegment = null;
+    if (indexSegmentIterator.hasNext()) {
+      indexSegment = indexSegmentIterator.next();
+      validDataSize += getValidDataSizePerIndexSegment(indexSegment, referenceTime);
+    }
+    while (indexSegmentIterator.hasNext()) {
+      IndexSegment nextIndexSegment = indexSegmentIterator.next();
+      if (nextIndexSegment.getLogSegmentName().equals(indexSegment.getLogSegmentName())) {
+        // index segment refers to same log segment as previous index segment
+        validDataSize += getValidDataSizePerIndexSegment(nextIndexSegment, referenceTime);
+      } else {
+        // index segment refers to a new log segment compared to previous index segment
+        String logSegmentName = indexSegment.getLogSegmentName();
+        totalValidDataSize += validDataSize;
+        if (validDataSizeBySegments != null) {
+          validDataSizeBySegments.put(logSegmentName, validDataSize);
+        }
+        indexSegment = nextIndexSegment;
+        validDataSize = getValidDataSizePerIndexSegment(nextIndexSegment, referenceTime);
+      }
+    }
+    if (index.indexes.size() > 1) {
+      String logSegmentName = indexSegment.getLogSegmentName();
+      totalValidDataSize += validDataSize;
+      if (validDataSizeBySegments != null) {
+        validDataSizeBySegments.put(logSegmentName, validDataSize);
+      }
+    }
+    return new Pair<>(referenceTime, totalValidDataSize);
+  }
+
+  /**
+   * Get valid data size of an index segment
+   * @param indexSegment the {@link IndexSegment} for which valid data size has to be determined
+   * @param referenceTimeInMs time in ms used as reference to check for expiration
+   * @return the valid data size of the given index segment
+   * @throws IOException
+   * @throws StoreException
+   */
+  private long getValidDataSizePerIndexSegment(IndexSegment indexSegment, long referenceTimeInMs)
+      throws IOException, StoreException {
+    List<MessageInfo> messageInfos = new ArrayList<>();
+    indexSegment.getEntriesSince(null, new FindEntriesCondition(Integer.MAX_VALUE), messageInfos, new AtomicLong(0));
+    long validSize = 0;
+    for (MessageInfo messageInfo : messageInfos) {
+      IndexValue value = index.findKey(messageInfo.getStoreKey());
+      if (!PersistentIndex.isDeleted(value) && !isExpired(messageInfo.getExpirationTimeInMs(), referenceTimeInMs)) {
+        validSize += messageInfo.getSize();
+      }
+    }
+    return validSize;
+  }
+
+  /**
+   * Check if {@code expirationTimeInMs} has expired compared to {@code referenceTimeInMs}
+   * @param expirationTimeInMs time in ms to be checked for expiration
+   * @param referenceTimeInMs the epoch time to use to check for expiration
+   * @return {@code true} if {@code expirationTimeInMs} expired wrt {@code referenceTimeInMs}, {@code false} otherwise
+   */
+  private boolean isExpired(long expirationTimeInMs, long referenceTimeInMs) {
```

Comment:
Feel this method is unnecessary. Wouldn't you always want to check against `now`?